### PR TITLE
feature-erms-3600

### PR DESCRIPTION
**Pull request by croqueGrec09**
as of ERMS-3600, a cleanup SQL function has been implemented which merges orgs given two keys

### CODE REVIEW

File: grails-app/migrations/functions/org_merge_function.sql [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+		loop
+		delete from org_setting where os_key_enum = osrow4.os_key_enum and os_org_fk = old_key;
+		end loop;
+	update org_setting set os_org_fk = new_key where os_org_fk = old_key;
+	delete from task WHERE tsk_org_fk = old_key and tsk_creator_fk in (select tsk_creator_fk from task where tsk_org_fk = new_key);
+	update task set tsk_org_fk = new_key where tsk_org_fk = old_key;
+	delete from org_type where org_id = old_key;
+	delete from person_role WHERE pr_org_fk = old_key and pr_prs_fk in (select pr_prs_fk from person_role where pr_org_fk = new_key);
+	update person_role set pr_org_fk = new_key where pr_org_fk = old_key;
+	delete from person WHERE prs_tenant_fk = old_key and prs_last_name in (select prs_last_name from person where prs_tenant_fk = new_key);
```
Comment (klodav) [inline grails-app/migrations/functions/org_merge_function.sql:124]: Max Müller != Michael Müller

File: grails-app/migrations/functions/org_merge_function.sql [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+		end loop;
+	for osrow4 in select os_key_enum from org_setting where os_org_fk = new_key and os_string_value is null and os_rv_fk is null and os_role_fk is null
+		loop
+		delete from org_setting where os_key_enum = osrow4.os_key_enum and os_org_fk = old_key;
+		end loop;
+	update org_setting set os_org_fk = new_key where os_org_fk = old_key;
+	delete from task WHERE tsk_org_fk = old_key and tsk_creator_fk in (select tsk_creator_fk from task where tsk_org_fk = new_key);
+	update task set tsk_org_fk = new_key where tsk_org_fk = old_key;
+	delete from org_type where org_id = old_key;
+	delete from person_role WHERE pr_org_fk = old_key and pr_prs_fk in (select pr_prs_fk from person_role where pr_org_fk = new_key);
```
Comment (klodav) [inline grails-app/migrations/functions/org_merge_function.sql:122]: loss of roles (function, responsibility, position) ?

Comment (croqueGrec09) [reply grails-app/migrations/functions/org_merge_function.sql:122]: Why? I check against the person itself which is linked to both orgs ...?

File: grails-app/migrations/functions/org_merge_function.sql [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+		end loop;
+	select * into oldorgdata from org where org_id = old_key;
+	update org set org_shortname = oldorgdata.org_shortname where org_id = new_key and org_shortname is null;
+	update org set org_url = oldorgdata.org_url where org_id = new_key and org_url is null;
+    delete from contact WHERE ct_org_fk = old_key and ct_content in (select ct_content from contact where ct_org_fk = new_key);
+	update contact set ct_org_fk = new_key where ct_org_fk = old_key;
+	delete from doc_context WHERE dc_org_fk = old_key and dc_doc_fk in (select dc_doc_fk from contact where dc_org_fk = new_key);
+	update doc_context set dc_org_fk = new_key where dc_org_fk = old_key;
+	update doc_context set dc_target_org_fk = new_key where dc_target_org_fk = old_key;
+	delete from identifier WHERE id_org_fk = old_key and id_value in (select id_value from identifier where id_org_fk = new_key);
```
Comment (klodav) [inline grails-app/migrations/functions/org_merge_function.sql:49]: loss of identifiers - EISSN:123 != PISSN: 123

File: grails-app/migrations/functions/org_merge_function.sql [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+	update identifier set id_org_fk = new_key where id_org_fk = old_key;
+	for oprow1 in select op_string_value, op_type_fk from org_property where op_owner_fk = new_key
```
Comment (klodav) [inline grails-app/migrations/functions/org_merge_function.sql:51]: loss of properties - mixing property types and data values

File: grails-app/migrations/functions/org_merge_function.sql [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+		loop
+		delete from org_property where op_string_value = oprow1.op_string_value and op_type_fk = oprow1.op_type_fk and op_owner_fk = old_key;
+		end loop;
+	for oprow2 in select op_date_value, op_type_fk from org_property where op_owner_fk = new_key
```
Comment (klodav) [inline grails-app/migrations/functions/org_merge_function.sql:55]: loss of properties - mixing property types and data values

File: grails-app/migrations/functions/org_merge_function.sql [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+		loop
+		delete from org_property where op_date_value = oprow2.op_date_value and op_type_fk = oprow2.op_type_fk and op_owner_fk = old_key;
+		end loop;
+	for oprow3 in select op_dec_value, op_type_fk from org_property where op_owner_fk = new_key
```
Comment (klodav) [inline grails-app/migrations/functions/org_merge_function.sql:59]: loss of properties - mixing property types and data values

File: grails-app/migrations/functions/org_merge_function.sql [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+		loop
+		delete from org_property where op_dec_value = oprow3.op_dec_value and op_type_fk = oprow3.op_type_fk and op_owner_fk = old_key;
+		end loop;
+	for oprow4 in select op_int_value, op_type_fk from org_property where op_owner_fk = new_key
```
Comment (klodav) [inline grails-app/migrations/functions/org_merge_function.sql:63]: loss of properties - mixing property types and data values

File: grails-app/migrations/functions/org_merge_function.sql [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+		loop
+		delete from org_property where op_int_value = oprow4.op_int_value and op_type_fk = oprow4.op_type_fk and op_owner_fk = old_key;
+		end loop;
+	for oprow5 in select op_ref_value_rv_fk, op_type_fk from org_property where op_owner_fk = new_key
```
Comment (klodav) [inline grails-app/migrations/functions/org_merge_function.sql:67]: loss of properties - mixing property types and data values

File: grails-app/migrations/functions/org_merge_function.sql [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+		loop
+		delete from org_property where op_ref_value_rv_fk = oprow5.op_ref_value_rv_fk and op_type_fk = oprow5.op_type_fk and op_owner_fk = old_key;
+		end loop;
+	for oprow6 in select op_url_value, op_type_fk from org_property where op_owner_fk = new_key
```
Comment (klodav) [inline grails-app/migrations/functions/org_merge_function.sql:71]: loss of properties - mixing property types and data values

File: grails-app/migrations/functions/org_merge_function.sql [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+			loop
+			delete from org_role where or_tipp_fk = orrow3.or_tipp_fk and or_org_fk = old_key;
+			end loop;
+		for orrow4 in select or_pkg_fk from org_role where or_org_fk = new_key
+			loop
+			delete from org_role where or_pkg_fk = orrow4.or_pkg_fk and or_org_fk = old_key;
+			end loop;
+		update org_role set or_org_fk = new_key where or_org_fk = old_key;
+	end if;
+	for osrow1 in select os_key_enum, os_string_value from org_setting where os_org_fk = new_key
```
Comment (klodav) [inline grails-app/migrations/functions/org_merge_function.sql:102]: loss of settings - mixing types and data values

File: grails-app/migrations/functions/org_merge_function.sql [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+		loop
+		delete from org_setting where os_string_value = osrow1.os_string_value and os_key_enum = osrow1.os_key_enum and os_org_fk = old_key;
+		end loop;
+	for osrow2 in select os_key_enum, os_rv_fk from org_setting where os_org_fk = new_key
```
Comment (klodav) [inline grails-app/migrations/functions/org_merge_function.sql:106]: loss of settings - mixing types and data values

File: grails-app/migrations/functions/org_merge_function.sql [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+		loop
+		delete from org_setting where os_rv_fk = osrow2.os_rv_fk and os_key_enum = osrow2.os_key_enum and os_org_fk = old_key;
+		end loop;
+	for osrow3 in select os_key_enum, os_role_fk from org_setting where os_org_fk = new_key
```
Comment (klodav) [inline grails-app/migrations/functions/org_merge_function.sql:110]: loss of settings - mixing types and data values

File: grails-app/migrations/functions/org_merge_function.sql [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+		loop
+		delete from org_setting where os_role_fk = osrow3.os_role_fk and os_key_enum = osrow3.os_key_enum and os_org_fk = old_key;
+		end loop;
+	for osrow4 in select os_key_enum from org_setting where os_org_fk = new_key and os_string_value is null and os_rv_fk is null and os_role_fk is null
```
Comment (klodav) [inline grails-app/migrations/functions/org_merge_function.sql:114]: loss of settings - mixing types and data values

File: grails-app/migrations/functions/org_merge_function.sql [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+			delete from org_role where or_shared_from_fk = orrow2.or_id; --aiming one certain org_role object
+			update org_role set or_shared_from_fk = orrow2.or_id where or_sub_fk in (select sub_id from subscription where sub_parent_sub_fk = orrow2.or_sub_fk) and or_org_fk = old_key;
+			update org_role set or_is_shared = true where or_id = orrow2.or_id and exists (select or_id from org_role where or_shared_from_fk = orrow2.or_id);
+			delete from org_role where or_sub_fk = orrow2.or_sub_fk and or_org_fk = old_key;
+			end loop;
+		for orrow3 in select or_tipp_fk from org_role where or_org_fk = new_key
+			loop
+			delete from org_role where or_tipp_fk = orrow3.or_tipp_fk and or_org_fk = old_key;
+			end loop;
+		for orrow4 in select or_pkg_fk from org_role where or_org_fk = new_key
```
Comment (klodav) [inline grails-app/migrations/functions/org_merge_function.sql:96]: loss of data if fk is null

File: grails-app/migrations/functions/org_merge_function.sql [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+			delete from org_role where or_lic_fk = orrow1.or_lic_fk and or_org_fk = old_key;
+			end loop;
+		for orrow2 in select or_id, or_sub_fk from org_role where or_org_fk = new_key
+			loop
+			delete from org_role where or_shared_from_fk = orrow2.or_id; --aiming one certain org_role object
+			update org_role set or_shared_from_fk = orrow2.or_id where or_sub_fk in (select sub_id from subscription where sub_parent_sub_fk = orrow2.or_sub_fk) and or_org_fk = old_key;
+			update org_role set or_is_shared = true where or_id = orrow2.or_id and exists (select or_id from org_role where or_shared_from_fk = orrow2.or_id);
+			delete from org_role where or_sub_fk = orrow2.or_sub_fk and or_org_fk = old_key;
+			end loop;
+		for orrow3 in select or_tipp_fk from org_role where or_org_fk = new_key
```
Comment (klodav) [inline grails-app/migrations/functions/org_merge_function.sql:92]: loss of data if fk is null

File: grails-app/migrations/org_merge_function.sql [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+	for altname in select org_name from org where org_id = old_key
+		loop
+		insert into alternative_name (altname_version, altname_name, altname_org_fk) values (0, altname, new_key);
+		end loop;
+	select * into oldorgdata from org where org_id = old_key;
+	update org set org_shortname = oldorgdata.org_shortname where org_id = new_key and org_shortname is null;
+	update org set org_url = oldorgdata.org_url where org_id = new_key and org_url is null;
+    delete from contact WHERE ct_org_fk = old_key and ct_content in (select ct_content from contact where ct_org_fk = new_key);
+	update contact set ct_org_fk = new_key where ct_org_fk = old_key;
+	delete from doc_context WHERE dc_org_fk = old_key and dc_doc_fk in (select dc_doc_fk from contact where dc_org_fk = new_key);
```
Comment (klodav) [inline grails-app/migrations/org_merge_function.sql:52]: select dc_doc_fk from contact [..]

**Review by klodav — APPROVED** [review]

**Pull request merged**